core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java

```diff
@@ -297,15 +330,27 @@ public static long rescale(long value, int fromScale, int toScale)
         if (toScale < fromScale) {
             throw new IllegalArgumentException("target scale must be larger than source scale");
         }
-        return value * longTenToNth(toScale - fromScale);
+
+        // To not overflow, abs(val) * pow(10, m) < pow(10, MAX_SHORT_PRECISION) => abs(val) < pow(10, MAX_SHORT_PRECISION - m)
+        if (toScale <= MAX_SHORT_PRECISION && abs(value) < longTenToNth(MAX_SHORT_PRECISION - (toScale - fromScale))) {
```

Comment:
I still wonder what the perf penality for additional check is and if it's worth it. Opinions @findepi @losipiuk @kokosing ?

Reply:
Can we have a micro-benchmark to check if that actually is a noticable penality?

Reply:
Do we have any docs or pointers regarding how to do it?

Reply:
You could take a look at: `io.trino.type.BenchmarkDecimalOperators` or `io.trino.orc.BenchmarkOrcDecimalReader`. I think the latter would be more appropriate (to get e2e difference)

The reason I wonder if it makes sense to validate numbers is that I assume connectors work, so if perf penalty is non-negligible we should skip it.

Reply:
Makes sense because I did not see any issues raised related to out of bounds decimals, so connectors are handling this validation somehow at their own individual level

Reply:
I have temporarily added the Benchmark code so that it's easier to get it reviewed. Will remove it once we are done with benchmarking.

Reply:
master - af9feda014b2cab18f2a294aeb2c07f0a7495829
current branch - aa8dbce9181a6e518e09b1eea53ef2b026420f84

Master benchmark results :
```bash
Result "io.trino.orc.BenchmarkOrcDecimalRescaleReader.readDecimal":
  0.013 ±(99.9%) 0.001 s/op [Average]
  (min, avg, max) = (0.013, 0.013, 0.014), stdev = 0.001
  CI (99.9%): [0.013, 0.013] (assumes normal distribution)


# Run complete. Total time: 00:01:10

Benchmark                                     Mode  Cnt  Score    Error  Units
BenchmarkOrcDecimalRescaleReader.readDecimal  avgt   60  0.013 ±  0.001   s/op
```

PR branch benchmark :
```bash
Result "io.trino.orc.BenchmarkOrcDecimalRescaleReader.readDecimal":
  0.015 ±(99.9%) 0.001 s/op [Average]
  (min, avg, max) = (0.015, 0.015, 0.016), stdev = 0.001
  CI (99.9%): [0.015, 0.015] (assumes normal distribution)


# Run complete. Total time: 00:01:10

Benchmark                                     Mode  Cnt  Score    Error  Units
BenchmarkOrcDecimalRescaleReader.readDecimal  avgt   60  0.015 ±  0.001   s/op
[master-branch-benchmark-output.txt](https://github.com/trinodb/trino/files/6327904/master-branch-benchmark-output.txt)
[orc-decimal-evolution-branch-benchmark.txt](https://github.com/trinodb/trino/files/6327905/orc-decimal-evolution-branch-benchmark.txt)


```

Reply:
[Master branch benchmark run](https://github.com/trinodb/trino/files/6327908/master-branch-benchmark-output.txt)
[PR branch benchmark run](https://github.com/trinodb/trino/files/6327910/orc-decimal-evolution-branch-benchmark.txt)

Reply:
Please ignore the benchmark Java file added right now. I will remove it once our discussion on bench marking is complete.